File: rust/src/dense.rs
use crate::perceptron::{Perceptron, ActivationType};

pub struct Dense {
    hidden_layer: Vec<Perceptron>,
    out_layer: Vec<Perceptron>,
}

impl Dense {
    pub fn new(
        nb_input: usize,
        nb_hidden: usize,
        hidden_act: ActivationType,
        nb_output: usize,
        output_act: ActivationType,
    ) -> Self {
        let hidden_layer = (0..nb_hidden)
            .map(|_| Perceptron::new(nb_input, hidden_act.clone()))
            .collect();

        let out_layer = (0..nb_output)
            .map(|_| Perceptron::new(nb_hidden, output_act.clone()))
            .collect();

        Dense {
            hidden_layer,
            out_layer,
        }
    }

    pub fn forward(&mut self, inputs: Vec<f64>) -> Vec<f64> {
        let hidden_outputs: Vec<f64> = self
            .hidden_layer
            .iter_mut()
            .map(|perceptron| perceptron.forward(&inputs))
            .collect();

        let final_outputs: Vec<f64> = self
            .out_layer
            .iter_mut()
            .map(|perceptron| perceptron.forward(&hidden_outputs))
            .collect();

        final_outputs
    }
}

File: python/dense.py
from typing import List
from perceptron import Perceptron, ActivationType


class Dense:
    def __init__(
        self,
        nbInput: int,
        nbHidden: int,
        hiddenAct: ActivationType,
        nbOutput: int,
        outputAct: ActivationType,
    ):
        self.hiddenLayer: List[Perceptron] = [
            Perceptron(nbInput, hiddenAct) for _ in range(nbHidden)
        ]
        self.outLayer: List[Perceptron] = [
            Perceptron(nbHidden, outputAct) for _ in range(nbOutput)
        ]
        self.outLayer: List[Perceptron] = [
            Perceptron(nbHidden, outputAct) for _ in range(nbOutput)
        ]

    def forward(self, inputs: List[float]) -> List[float]:
        hidden_output = [
            perceptron.forward(inputs) for perceptron in self.hiddenLayer
        ]
        return [
            perceptron.forward(hidden_output) for perceptron in self.outLayer
        ]

    def backward(self, deltas: List[float]):
        out_perceptron_delta = self.outLayer[0].backward(deltas[0])

        for perceptron in self.hiddenLayer:
            perceptron.backward(out_perceptron_delta)
